versioning/create-release.py

```diff
@@ -0,0 +1,191 @@
+    try:
+        f = open(version_file, "w")
+        f.write(ver_str)
+        f.close()
+    except:
+        error_exit('unable to write version to file')
+
+# wrapper function for calling git commands
+def gitcmd(command, error_string):
+    try:
+        res = git(command.split(' ')).strip('\n')
+    except:
+        error_exit(error_string)
+    return res
+
+# check for untracked files or uncommitted changes
+def repo_is_clean():
+    res = gitcmd('status --porcelain', 'git status failed')
+    if res != '':
+        error_exit('uncommitted changes or untracked files present')
+
+# check if current branch matches the requested branch
+def check_branch(branch):
+    res = gitcmd('rev-parse --abbrev-ref HEAD', 'git rev-parse failed')
+    if (res != branch):
+        error_exit('current branch {} != {}'.format(res, branch))
+
+def git_checkout_branch(branch):
+    gitcmd('checkout {}'.format(branch), 'unable to checkout branch {}'.format(branch))
+
+def git_checkout_create_branch(branch):
+    res = gitcmd('checkout -b {}'.format(branch), 'unable to checkout/create branch {}'.format(branch))
+    print(res)
+
+# Return tags as a string, replace newlines with space
+def git_get_tags():
+    return gitcmd('tag', 'failed to get tags').replace('\n', ' ')
+
+
+# used for both tags and VERSION file, strips leading 'v'
+# 'v1.2.3' -> [1, 2, 3] and '1.2.3' -> [1, 2, 3]
+def version_from_string(ver_str):
+    if len(ver_str) < 5:
+        return [0, 0, 0]
+    if ver_str[0] == 'v':
+        ver_str = ver_str[1:]
+    res = ver_str.split('.')
+    if len(res) != 3:
+        return [0, 0, 0]
+    return [int(res[0]), int(res[1]), int(res[2])]
+
+def latest_ver_from_file():
+    try:
+        res = cat(version_file).strip('\n')
+    except:
+        error_exit('failed to get version from file')
+    return version_from_string(res)
+
+def latest_ver_from_list(versions):
+    curmaj, curmin, curpat = [0, 0, 0]
+    for v in versions.split(' '):
+        maj, min, pat = version_from_string(v.replace(' ', ''))
+        if (maj > curmaj) or (maj == curmaj and min > curmin) or (maj == curmaj and min == curmin and pat > curpat):
+            curmaj, curmin, curpat = [ maj, min, pat]
+    return [curmaj, curmin, curpat]
+
+# return version strings from version numbers
+def strings_from_version(ver):
+    version = '{}.{}.{}'.format(ver[0], ver[1], ver[2])
+    tag = 'v'+version
+    branch  = 'release-{}.{}'.format(ver[0], ver[1])
+    return [version, tag, branch]
+
+# increment major, minor or patch numbers, return new version strings
+def bump_version(ver, release):
+    if release == "major":
+        new = [ver[0]+1, 0, 0]
+    elif release == "minor":
+        new = [ver[0], ver[1]+1, 0]
+    elif release == "patch":
+        new = [ver[0], ver[1], ver[2]+1]
+    else:
+        error_exit('Illegal relase (use major/minor/patch)')
+    return strings_from_version(new)
+
+#
+# command line options
+parser = argparse.ArgumentParser()
```

Comment:
I usually put the "main" execution code in a `main()` function. This code is then executed by placing the following two lines somewhere in the file:
```
if __name__ == '__main__':
    main()
```
I think this makes the code easier to read and easier to edit.

Reply:
agree, will do

Reply:
done